extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/ThreadLocalPool.java

```diff
@@ -36,26 +56,41 @@ public ThreadLocalPool(Vertx vertx, PoolOptions poolOptions) {
         this.poolOptions = poolOptions;
     }
 
-    private PoolType pool() {
+    PoolType pool() {
         checkPoolIsOpen();
         PoolType pool = threadLocal.get();
         if (pool == null) {
-            synchronized (threadLocalPools) {
+            synchronized (allConnections) {
                 checkPoolIsOpen();
                 pool = createThreadLocalPool();
-                threadLocalPools.add(pool);
+                allConnections.add(new PoolAndThread(pool));
                 threadLocal.set(pool);
+                scanForAbandonedConnections();
             }
         }
         return pool;
     }
 
+    private final void scanForAbandonedConnections() {
+        for (PoolAndThread pair : allConnections) {
+            if (pair.isDead()) {
+                //This might potentially close the connection a second time,
+                //so we need to ensure implementations allow it.
+                pair.close();
```

Comment:
This looks like it doesn't remove the entry from `allConnections`, does it?

Reply:
it does ;)

Reply:
it's also being tested.

Reply:
Oh, it does, via the `Pool` wrapper. OK thanks.

Reply:
yea, sorry it's horribly convoluted. I would have preferred to keep all responsibilities within this single class, but then I wouldn't have been able to expose the various flavours of different types the subclasses have to produce.

I do wonder if we shouldn't rather expose a more limited interface than each of `PgPool`, `MySQLPool`.. it's kinda weird to expose the pool lifecycle methods, but I'd not break the current API when vert.x 4 is around the corner.